Fix Issue #5050

alextremblay opened the pull request:
<!--
    Thank you for your interest in contributing to Nautobot! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->
# Closes #5050
# What's Changed
<!--
    Please include:
    - A summary of the proposed changes
    - A sectioned breakdown for larger features under ## subheadings
-->
updated the `available_on_device`/`get_for_device` filter in the `VLANFilterSet` to include not just the device(s)' locations, but also those location(s)' ancestors
# Screenshots
<!--
    - Screenshots, example payloads where relevant:
      - Before/After for bugfixes
      - Using a new feature
-->
With vlans assigned to a location (a building) and a switch assigned to a sub-location (a room), here's the before and after

Before:
![image](https://github.com/user-attachments/assets/b7c46fad-91c1-4fcb-9af4-26264328f606)

After:
![image](https://github.com/user-attachments/assets/ad588ee7-7707-4e77-87f2-7d7a01888688)

# TODO
<!--
    Please feel free to update todos to keep track of your own notes for WIP PRs.
-->
- [x] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

HanlinMiao commented:
We will implement some test cases for this fix and examine the performance implications of calling `location.ancestors()`.

HanlinMiao commented:
Hi @alextremblay, thank you for your contribution.
Here is the PR https://github.com/nautobot/nautobot/pull/6195 with your changes and some unit tests built as well. Feel free to take a look and add the unit tests to your branch since I seem to not have write permission to your forked repo.

glennmatthews commented on changes/6091.fixed:
Can remove this file, `5050.fixed` is sufficient.

glennmatthews commented on changes/5050.fixed:
```diff
@@ -0,0 +1 @@
+Allow VLANs assigned to a device's location's parent locations (incl. parents of parents etc) to be assigned to that device's interfaces
```
```suggestion
Changed logic to permit VLANs assigned to a device's location's parent locations (including parents of parents, etc.) to be assigned to that device's interfaces.
```

glennmatthews commented on nautobot/dcim/api/serializers.py:
```diff
@@ -797,13 +797,19 @@ class Meta:
+        if device:
+            location = device.location
+        if location:
+            location_ids = [ancestor.id for ancestor in location.ancestors()]
+            location_ids.append(location.id)
```
```suggestion
            location_ids = location.ancestors(include_self=True).values_list("id", flat=True)
```

glennmatthews replied:
It occurs to me we should probably add an `.ancestor_ids()` method to TreeQuerySet as well, as this is a bit recursive - `.ancestors()` gets the list of PKs itself and then uses that list to derive a queryset, but here we then use the queryset to get the list of PKs. Not blocking for this PR though.

glennmatthews commented on nautobot/dcim/forms.py:
```diff
@@ -213,16 +213,20 @@ def clean(self):
-            valid_location = self.cleaned_data[parent_field].location
+            location = self.cleaned_data[parent_field].location
+            if location:
+                location_ids = [ancestor.id for ancestor in location.ancestors()]
+                location_ids.append(location.id)
```
```suggestion
                location_ids = location.ancestors(include_self=True).values_list("id", flat=True)
```

glennmatthews commented on nautobot/dcim/forms.py:
```diff
@@ -213,16 +213,20 @@ def clean(self):
+                and not VLANLocationAssignment.objects.filter(location__in=location_ids, vlan=v).exists()
             ]
 
             if invalid_vlans:
```
Mind updating the below `ValidationError` text similarly to what you did in `serializers.py` above?

glennmatthews commented on nautobot/dcim/models/device_components.py:
```diff
@@ -727,13 +727,19 @@ def clean(self):
+        if self.parent:
+            location = self.parent.location
+        if location:
+            location_ids = [ancestor.id for ancestor in location.ancestors()]
+            location_ids.append(location.id)
```
```suggestion
            location_ids = location.ancestors(include_self=True).values_list("id", flat=True)
```

glennmatthews commented on nautobot/dcim/models/device_components.py:
```diff
@@ -727,13 +727,19 @@ def clean(self):
+            and not self.untagged_vlan.locations.filter(pk__in=location_ids).exists()
         ):
             raise ValidationError(
                 {
```
Mind updating this ValidationError as well?

glennmatthews commented on nautobot/dcim/tests/test_forms.py:
```diff
@@ -343,9 +343,20 @@ def setUpTestData(cls):
         """Assert that form validation fails when no matching locations are associated to tagged VLAN"""
-        self.vlan.locations.set(list(Location.objects.exclude(pk=self.device.location.pk))[:2])
+        location = self.device.location
+        location_ids = [ancestor.id for ancestor in location.ancestors()]
+        self.vlan.locations.set(list(Location.objects.exclude(pk__in=location_ids))[:2])
```
Intentional that we're *not* excluding `self.device.location` from this list any more?

glennmatthews commented on nautobot/dcim/tests/test_models.py:
```diff
@@ -2378,19 +2378,58 @@ def test_tagged_vlan_raise_error_if_mode_not_set_to_tagged(self):
+        intf_role = Role.objects.get_for_model(Interface).first()
+        location_type = LocationType.objects.get(name="Campus")
+        parent_location = Location.objects.filter(children__isnull=False, location_type=location_type).first()
+        child_location = Location.objects.filter(parent=parent_location, location_type=location_type).first()
```
does `child_location` *need* to be a `Campus` `location_type`, or is any child of `parent_location` sufficient?

glennmatthews commented on nautobot/dcim/utils.py:
```diff
@@ -55,14 +55,11 @@ def cable_status_color_css(record):
-            ColorChoices.COLOR_AMBER: "warning",
-            ColorChoices.COLOR_CYAN: "info",
-        }
-        status_color = record.cable.get_status_color().strip("#")
-        return CABLE_STATUS_TO_CSS_CLASS.get(status_color, "")
+    # The status colors are for use with labels and such, and tend to be quite bright.
+    # For this function we want a much milder, mellower color suitable as a row background.
+    base_color = record.cable.get_status_color().strip("#")
+    lighter_color = rgb_to_hex(*lighten_color(*hex_to_rgb(base_color), 0.75))
+    return f"background-color: #{lighter_color}"
```
Unintended change? Bad merge perhaps?

glennmatthews commented on nautobot/dcim/utils.py:
```diff
@@ -139,20 +136,23 @@ def validate_interface_tagged_vlans(instance, model, pk_set):
+    # a location that is not the parent's location, or parent's location's ancestors, or None
+    location = getattr(instance.parent, "location", None)
+    if location:
+        location_ids = [ancestor.id for ancestor in location.ancestors()]
+        location_ids.append(location.id)
```
```suggestion
        location_ids = location.ancestors(include_self=True).values_list("id", flat=True)
```

glennmatthews commented on nautobot/extras/tests/test_forms.py:
Bad merge?

glennmatthews commented on output.txt:
Remove?

alextremblay commented:
Hi @glennmatthews, thanks for the review! 

I realize now that simply merging the changes from Hanlin's branch was a mistake

Cherry-picking the unittest-related commits also didn't work, since those commits included extraneous changes not related to this PR.

In the end, I rebased my branch of of the latest development commit, re-implemented the unit tests suggested by Hanlin, and implemented the changes you suggested in your code review.

The PR should now be much cleaner and clearer, easier to review

glennmatthews commented on changes/6091.fixed:
```diff
@@ -0,0 +1 @@
+Allow VLANs assigned to a device's location's parent locations (incl. parents of parents etc) to be assigned to that device's interfaces
```
Remove file?

glennmatthews commented on nautobot/dcim/api/serializers.py:
```diff
@@ -797,10 +797,13 @@ class Meta:
+            if vlan.locations.exists() and not vlan.locations.filter(pk__in=location_ids).exists():
                 raise serializers.ValidationError(
                     {
                         "tagged_vlans": f"VLAN {vlan} must have a common location as the interface's parent device, or "
```
Update ValidationError text here?

glennmatthews commented on nautobot/dcim/utils.py:
```diff
@@ -139,12 +139,14 @@ def validate_interface_tagged_vlans(instance, model, pk_set):
+        model.objects.filter(pk__in=pk_set).exclude(locations__isnull=True).exclude(locations__in=location_ids)
     )
 
     if tagged_vlans.count():
```
Update ValidationError text here?

glennmatthews commented on nautobot/dcim/models/device_components.py:
```diff
@@ -727,13 +727,17 @@ def clean(self):
         # Validate untagged VLAN
-        # TODO: after Location model replaced Site, which was not a hierarchical model, should we allow users to assign a VLAN belongs to
-        # the parent Locations or the child locations of `device.location`?
+        location = self.parent.location
```
`self.parent` might be `None`, so perhaps:

```suggestion
        location = self.parent.location if self.parent is not None else None
```

HanlinMiao commented:
Closed by https://github.com/nautobot/nautobot/pull/6379
